bqskit/runtime/__init__.py

```diff
@@ -0,0 +1,183 @@
+the end user. See :class:`~bqskit.runtime.attached.AttachedServer` for
+more info.
+
+A server running in attached mode is limited to a single machine. If you
+would like to parallelize BQSKit across multiple nodes over a network,
+you can launch a server independently in detached mode. Here, you first
+start :class:`~bqskit.runtime.manager.Manager` processes on all machines.
+Then link them all together by running a
+:class:`~bqskit.runtime.detached.DetachedServer` process, which Compilers'
+can connect to by passing the appropriate address in their constructor.
+
+Upon installing BQSKit, two shell scripts are added to your environment
+that can spawn managers and detached servers::
+
+    bqskit-manager
+
+and::
+
+    bqskit-server <address_of_bqskit_managers>
+
+Typically, you start managers first on all nodes in the desired cluster.
+Then you can start the server with a comma seperated list of all managers
+ip address and optionally ports. You can see the `-h` option of each
+command or the :func:`~bqskit.runtime.detached.start_server` and
+:func:`~bqskit.runtime.manager.start_manager` entry points for
+more information.
+
+For more information on how to manage jobs submitted to a runtime server,
+see the :obj:`~bqskit.compiler` documentation.
+
+.. autosummary::
+    :toctree: autogen
+    :recursive:
+
+    ~bqskit.runtime.attached.AttachedServer
+    ~bqskit.runtime.detached.DetachedServer
+    ~bqskit.runtime.manager.Manager
+    ~bqskit.runtime.detached.start_server
+    ~bqskit.runtime.manager.start_manager
+
+Parallelize Pass Computation
+############################
+
+When developing a BQSKit compiler pass, you can hook into the active
+runtime server through the :func:`get_runtime` function. This returns a
+:class:`RuntimeHandle`, which you can use to submit, map, wait on, and
+cancel tasks in the execution environment.
+
+For more information on how to design a custom pass, see this (TODO, sorry,
+you can look at the source code of existing
+`passes <https://github.com/BQSKit/bqskit/blob/master/bqskit/passes>`_
+for a good example for the time being).
+
+.. autosummary::
+    :toctree: autogen
+    :recursive:
+
+    ~bqskit.runtime.get_runtime
+    ~bqskit.runtime.RuntimeHandle
+    ~bqskit.runtime.future.RuntimeFuture
+
+Standard Logging Interface
+##########################
+
+This system uses the standard python logging module to provide a familiar
+logging interface. On client processes -- the ones where you create the
+Compiler object -- you can configure python loggers like normal before
+submitting tasks and have the entire system honor that configuration.
+"""
+from __future__ import annotations
+
+import faulthandler
+import os
+from typing import Any
+from typing import Callable
+from typing import Protocol
+from typing import TYPE_CHECKING
+
+# Enable low-level fault handling: system crashes print a minimal trace.
+faulthandler.enable()
+
+
+# Disable multi-threading in BLAS libraries.
```

Comment:
Otherwise, when we spawn N workers, they each spawn N threads. I am leaving it to future work to control and analyze this better, i.e., N/k workers, k blas threads, with k configurable. For now, we just do N workers and 1 blas thread.